### DISCUSSION TITLE
Method of creating a component from scratch

### DISCUSSION THREAD

**Discussion by ThibFrgsGmz** [nasa/fprime#962] [Q&A] [answered]
Hi there,

What method do you use to create a new component from scratch?

Personally I copy and paste from an already developed component where I only keep the .xml files and start customizing them. This is clearly repetitive and prone to human error (name changes for example).

Do you have a better way to do this? Is there a fprime-util command that I don't know of to do this?

From my developer point of view, I would like to have available a "fprime-util create-comp" command to create a component.
This command would take several options like:

```shell
>$fprime-util create-comp --help

Create a file tree in a folder for the component to be created.

    -n, --name         name of the component
    -k, --kind         type of component (active, passive...)
```

This would generate a folder for the component containing the .xml template files for the type of component (active, passive...) desired and the cmake file for the build chain.

**Comment by SterlingPeet** [comment, marked as answer]
Hello @tThibFrgsGmz,

I built a cookiecutter template to make the process easier.  You can see the original standalone template here:

https://github.com/SterlingPeet/cookiecutter-fprime-component

However, I would recommend checking out the newer adapted version that is included in [fprime-tools](https://github.com/fprime-community/fprime-tools):

`fprime-util new -h`

See if that makes things easier for you.

Thanks,

Sterling Peet

**Reply by ThibFrgsGmz** [reply]
Indeed, it seems to meet the need I expressed! Although I've never used CookieCutter, from what I see of the commits, I get the impression that the work on the command is awesome.

Unfortunately, despite the ``-h`` helper in the command, at the moment I'm having trouble figuring out how to use it. I've tried several times and I always get errors, I must be using it wrong.

> Side note: do you have a reference to an educational resource that you would recommend to learn CookieCutter technology?

**Reply by Joshua-Anderson** [reply]
@ThibFrgsGmz Are you using `fprime-util new` or directly calling cookiecutter? If you're using `fprime-util new` I wouldn't expect to see any errors

**Reply by ThibFrgsGmz** [reply]
@Joshua-Anderson I just tested on my personal laptop and ``fprime-util new`` did not fail!

I'm having trouble seeing how to fill in the fields correctly but it should be fine with time and the documentation will improve I guess.

I see ``*ComponentImpl.cpp/hpp`` files being created, is this temporary or is there any real point?
I would expect that at this stage we are not yet concerned with how we implement the component.

I don't know why the command fails from my work computer.
Since I don't have access to the Internet from my office (we don't have a mirror server for the packages yet), I have to ask my IT department to update my fprime package.
I requested an update this Tuesday and so my version contains the ``fprime-util new`` command. 

Maybe the installation by my IT department went wrong or the command needs to access a resource on the internet?

**Reply by Joshua-Anderson** [reply]
> Maybe the installation by my IT department went wrong or the command needs to access a resource on the internet?

You're probably using the super-beta version of fprime-util new that downloads the cookiecutter template via github. The latest [fprime-tools version on github](https://github.com/fprime-community/fprime-tools), which will be part of the next pip release, uses a local template by default.

> I see *ComponentImpl.cpp/hpp files being created, is this temporary or is there any real point?

The goal is that you can run `fprime-util new` and then successfully build/compile your component right away. That way you can start adding functionality to your component while being able to compile the entire development process.
